docs/DMS/PRIMARY-DATA-STORAGE.md

```diff
@@ -25,3 +25,268 @@ Braindump:
   downstream systems. And they can use JSON Path.
 * Unlike Meadowlark, we probably need one table per entity.
+
+From 3/4/2024 Brainstorm:
+
+
+## Problems to solve via DB design
+
+### Existence and reference validation
+
+The first problem is existence and reference validation. Unlike Meadowlark, we want to leverage foreign key referential integrity to do this. Note we need to be able to take into account superclass identities, for example the existence of an EducationOrganization can satisfy reference validation for a School reference, so there has to be some mapping here. We would also like reference validation to be tunable in some fashion, meaning you can turn it on or off.
+
+### Query support in relational DB
+
+The next problem is query support. We'll need to know what the different fields are that you're allowed to query on, and then have some way to index them into the json.
+
+One question is how performant does this need to be? If/when it becomes too slow, moving to a search engine would be the recommendation because if you want true performance you would use a separate read only store.
+
+### Security
+
+The next thing that you need to be able to support is security. Following the ODS/API, we'll want Namespace-based for sure, Education Organization-based probably, and possibly Grade Level-based as well. In some ways this may be similar to query support, as we'll need to know the relevant fields to secure on for each resource.
+
+### Streaming changes
+
+We'll need the ability to stream out changes. This needs to be something that can be easily integrated with Kafka and potentially other streaming platforms, with a well-defined message format. Change query support is related to this, but we won't prejudge whether it will be based off of streaming.
+
+### Simplicity
+
+As maintainability is a primary goal, we want to keep Tanager as simple as possible. Meadowlark demonstrated that we can simplify over the ODS/API by avoiding code-per-resource and instead use a generalized design applicable to any resource. While it may not be possible with Tanager to do this completely, we would like to do it whenever possible.
+
+### Performance
+
+We need to make sure that we design to meet performance goals, and test early and often to ensure that we are able to hit those goals.
+
+## Potential Solutions
+
+### Option A - Similar to Meadowlark
+
+```mermaid
+erDiagram
+    References }o--|| Documents : ""
+    References {
+        bigint id PK "Auto-increment, otherwise unused"
+        Guid parent_ref_id FK "Parent document with reference"
+        Guid referenced_ref_id FK "Document being referenced"
+    }
+    Documents {
+        Guid referential_id PK "Hash of identity"
+        Guid document_uuid "API resource id, indexed"
+        string project_name "Example: Ed-Fi (for DS)"
+        string resource_name "Example: Student"
+        string resource_version "Example: 5.0.0"
+        JSON edfi_doc "The document"
```

Comment:
Good article: https://bornsql.ca/blog/think-twice-about-storing-json-in-your-sql-server-database/

Our use case is a bit different and the article's criticism don't necessarily apply. But they are good to think about.

Need to read the comments too. `COMPRESS` looks interesting.